sqlglot/optimizer/qualify_columns.py

```diff
@@ -213,6 +214,31 @@ def _qualify_columns(scope, resolver):
             # column_table can be a '' because bigquery unnest has no table alias
             if column_table:
                 column.set("table", column_table)
+        elif column_table not in scope.sources:
+            # structs are used like tables (e.g. "struct"."field"), so they need to be qualified
+            # separately by parsing them as <table>.<column>.dot.dot.dot ...
+
+            struct_fields = [val for val in reversed(list(column.args.values())) if val is not None]
+            struct_root = struct_fields[0]
+
+            if struct_root.name in scope.sources:
+                # struct is already qualified, but we still need to change the AST representation
+                struct_table = struct_root
+                struct_fields = struct_fields[1:]
+                struct_root = struct_fields[0]
+            else:
+                struct_table = resolver.get_table(struct_root.name)
+
+            if struct_table:
+                while column.parent and isinstance(column.parent, exp.Dot):
+                    column = column.parent
+                    struct_fields.append(column.expression)
+
+                new_column = exp.column(struct_root, table=struct_table)
```

Comment:
what if there's a db or catalog?

Reply:
I think the table here refers to the alias of the source `[catalog.][db.]table`. Can you provide an example & expected behavior so that I can check locally?

Reply:
Ok maybe there is a problem here:

```python
>>> from sqlglot.optimizer import optimize
>>>
>>> optimize("SELECT c.d.t.struct.f1.f2 AS f2 FROM c.d.t").sql()
'SELECT "c"."d"."t"."struct"."f1"."f2" AS "f2" FROM "c"."d"."t" AS "t"'
```

I guess we only want `"t"` here before `"struct"`, because of the alias?